{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "edbe3e02-a1ac-4540-b368-0a8b3d64c2a6",
   "metadata": {},
   "source": [
    "![image](https://cdn.discordapp.com/attachments/996200880351215636/1065002848355631165/New_Atlantis.png) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "321d56b0-ecb4-47e9-b471-8fcee9da8a87",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Execution of the antiSMASH Tool\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad9045a-b22a-4690-8566-cbf08e920b28",
   "metadata": {},
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3feb993-eb12-435d-a51f-856bc6028ea3",
   "metadata": {},
   "source": [
    " \n",
    " This notebook will give a demonstration of running the tool `antiSMASH` on assembled metagenomic data to identify Biosynthetic Gene Clusters (BGCs) from amongst the reads. `antiSMASH` stands for (antibiotics and Secondary Metabolite Analysis Shell) and is a comprehensive genome mining tool used for the automated identification, annotation, and analysis of secondary metabolite biosynthesis gene clusters in bacterial and fungal genomes. Secondary metabolites are bioactive compounds, often with therapeutic potential, including antibiotics, antifungals and antitumorals.\n",
    "'antiSMASH' has quite a few features, but here we utilize only its ability to identify BGCs It can identify all known types of BGCs such as those responsible for the production of polyketides, non-ribosomal peptides, terpenes, etc.\n",
    "   \n",
    "   \n",
    "The `antiSMASH` tool provides biologists and bioinformaticians with a useful means of deep-diving into genomic data to accelerate the discovery and diversification of biologically active secondary metabolites."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a41f9a9-6c2d-47a9-a063-5e9c5b07e0a5",
   "metadata": {},
   "source": [
    "**Markdown cell giving a brief explanation of the tool. Cover what insight is hoped to be generated by the tool and a extremely brief desciption of its methods.**\n",
    "\n",
    "Dont have a good grasp on why/how the tool works so please provide some brief comments on this."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77c9dea-7126-4093-8fb3-ef15b5a6494a",
   "metadata": {},
   "source": [
    "## Input Data \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e028d77b-b270-4032-949f-2d2bd47125e6",
   "metadata": {},
   "source": [
    "The input data are metagenomic samples previously preprocessed and assembled with [VEBA](https://github.com/jolespin/veba), a metagenomic assembly tool. For the purpose of this notebook, we will analyze metagenomes originally from the SOLA dataset. \n",
    "The input consists of:\n",
    "1. **Assembled contigs**: .fasta files.\n",
    "\n",
    "2. **Mapping files** of the assembled contigs in .bam files.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2803ae5-51dc-4a2d-b442-170a8f18412d",
   "metadata": {},
   "source": [
    "## Output Data \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b85d43d-7a3c-4b46-8059-6293d0ec700a",
   "metadata": {},
   "source": [
    "`antiSMASH` outputs an array of data related to the identified secondary metabolite biosynthesis gene clusters (BGCs). The output is highly comprehensive and includes the following:\n",
    "\n",
    "- **Gene Cluster Annotations:** For each identified BGC, antiSMASH provides an annotation of the genes in the cluster, including the predicted function and potential products. GenBank file.\n",
    "\n",
    "\n",
    "NOTE\n",
    "\n",
    "**got this from the internet so please verify accuracy and remove outputs that arent used for this purpose**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad90a79e-33d2-4b4e-bbfe-9987c18b25f4",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Environment\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf3371a7-5fe9-4258-a49b-05d599051e85",
   "metadata": {},
   "source": [
    "## Main Dependencies\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "493ca348-566c-400f-855e-13e2edb06cf2",
   "metadata": {},
   "source": [
    "The `antiSMASH` operating environment is fully decribed in the Dockerfile located in the src subfolder of this notebook. The docker image contains all necessary packages to run `antiSMASH`, all distributions are the most recent versions of their respective packages and distributed via apt. \n",
    "These packages include:\n",
    "\n",
    "- [hmmer 1/2](http://hmmer.org)\n",
    "- [diamond](https://github.com/bbuchfink/diamond)\n",
    "- [fasttree](http://www.microbesonline.org/fasttree/)\n",
    "- [prodigal](https://github.com/hyattpd/Prodigal)\n",
    "- [ncbi blast](https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html)\n",
    "- [muscle](https://www.ebi.ac.uk/Tools/msa/muscle/)\n",
    "- [glimmerhmm](https://ccb.jhu.edu/software/glimmerhmm/man.shtml)\n",
    "    \n",
    "The `antiSMASH` [distribution](https://antismash.secondarymetabolites.org) and its associated refernce data are pulled directly from their website.\n",
    "For more information and detail on how these packages are installed, see the Dockerfile.\n",
    "\n",
    "NOTE: Add pointer to image on dockerhub when uploaded."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8006bc65-d07b-47bc-936f-c081f40ccf0e",
   "metadata": {},
   "source": [
    "## Notebook Utility Installs\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f3df6f8-0e74-46bc-8d2e-06c810405d55",
   "metadata": {},
   "source": [
    "The import-ipynb package installed here provides utility in using a refernce notebook as a Python module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f21efd60-c031-4586-bb0a-23b552a60482",
   "metadata": {},
   "outputs": [],
   "source": [
    "#install for import-ipynb that will be applied later\n",
    "!pip install import-ipynb -q"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "424f73ed-af02-4075-8287-c38ca95356fd",
   "metadata": {},
   "source": [
    "## Pathing and Directory Environment Variables \n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d765ab46-5875-4a9d-97e5-4bcf21631ed8",
   "metadata": {},
   "source": [
    "**The following instructions set environment pathing variables and loads a jupyter extension for running R.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0135341f-6059-4f25-aecc-f8e1d27ae306",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: WORKDIR=data #working directory for executables\n",
      "env: seqtk=/nfs/bin/seqtk/seqtk #pathing for seqtk package\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython\n",
    "#working directory for executables \n",
    "%set_env WORKDIR=data \n",
    "\n",
    "#pathing for seqtk package\n",
    "%set_env seqtk=/nfs/bin/seqtk/seqtk #pathing for seqtk package\n",
    "\n",
    "#sets environ relatively in case files move, this sets the REPO variable to be the parent directory of this notebook.\n",
    "import os \n",
    "cwd = os.getcwd()\n",
    "parent_directory = os.path.dirname(cwd)\n",
    "\n",
    "#sets env variable \"REPO\" to name of parent directory\n",
    "os.environ['REPO'] = parent_directory #sets env variable \"REPO\" to name of parent directory\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "798a0f55-6f6f-4ef1-a7f1-bf72c9e9d57c",
   "metadata": {},
   "source": [
    "**These commands creates data and output directories if they don't already exist.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3c0d1c43-1e47-4e7f-9ba4-17380d069a68",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash \n",
    "mkdir -p ${WORKDIR}/data/sola\n",
    "mkdir -p ${WORKDIR}/outputs/antismash/taxonomy\n",
    "mkdir ${WORKDIR}/outputs/bgc_abund\n",
    "mkdir ${WORKDIR}/outputs/bgc_taxa\n",
    "mkdir ${WORKDIR}/outputs/tables\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba91e36d-a9db-4de2-a79e-597db5a0fced",
   "metadata": {},
   "source": [
    "## R Based Dependencies\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e00ac888-f20b-408a-aa81-f656cf927c61",
   "metadata": {},
   "source": [
    "Insert handling and instructions for the above."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f32a969e-f32d-4ff6-bd64-27244614a6c7",
   "metadata": {},
   "source": [
    "---\n",
    "## Import Statements (code)(import ipynb)\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "050d9b6c-a887-43bd-add9-263b913bfb53",
   "metadata": {},
   "outputs": [],
   "source": [
    "import import_ipynb \n",
    "#import all untility functions from utility_func nb \n",
    "from src.NB_utility import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e8a951-1ffa-4af3-a31f-9126426284c6",
   "metadata": {},
   "source": [
    "NOTE: your implementation does not appear to need this section, but that may change once we add features to the JHub so I am leaving it here for now."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e51905b-74dd-42d9-9b66-5e1e8f56cb60",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Parameters\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e23c198d-fcef-4acf-a53d-2d7af3cc9f07",
   "metadata": {},
   "source": [
    "Add handling for bash flags and r - object oriented parameter approach. parameter object?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3131b0ef-07d3-431b-9138-8c6f4e820496",
   "metadata": {},
   "source": [
    "This section should describe all parameters and setting used to input into the tool. All user supplied arguments should be defined and explained in this section.\n",
    "This section should alternate markdown and code, the markdown explains what a parameter does and what the options are. The first parameter cell should always be paths to data.\n",
    "The first section should contain one cell that only specifies data locations, such that a user only has to edit this cell in order to point the notebook to their data. The next parameter cell should contain paths to any config files or reference databases. The rest of the section should alternate between markdown explaining the parameter followed by code to set the parameter to some default value in either python or env variables. Duplicate the follwoing examples for each paramter used during data cleaning and execution of the tool."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0087c162-c5f3-43d7-a9a1-c74d17b53355",
   "metadata": {},
   "source": [
    "## Input and output data directories \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "197359c5-60ed-48a9-b171-f2e92e9dabc3",
   "metadata": {},
   "source": [
    "These variables are strings that conatin the path to the input and output data directories.\n",
    "\n",
    "This was set earlier, still adjusting the template but no need to worry about this section, but the previous directory structure works well."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bddb0e82-48da-49cc-8a66-dc8167d81558",
   "metadata": {},
   "source": [
    "## Run Time ENV Variables\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39d45434-2ace-41e3-8998-bdf9bc6d6826",
   "metadata": {},
   "source": [
    "### Example Sample ID\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62434768-bd53-4971-ad7e-f74f615fe99f",
   "metadata": {},
   "source": [
    "The ID of the sample being processed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e0bd006-519b-4bd6-a702-69455adc8898",
   "metadata": {},
   "outputs": [],
   "source": [
    "%set_env SAMPLE=ERR2604088"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5236bc20-abe6-464f-be53-5779700623cb",
   "metadata": {},
   "source": [
    "### Taxonomy of Focus \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e31180d9-47ab-463c-9760-839bc2470365",
   "metadata": {},
   "source": [
    "The specific taxonomy of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39ad029b-768c-46f8-b4cf-22710cbb1af5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%set_env TAXON=bacteria"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "664fe629-c123-46dc-8804-7c614e8cfee7",
   "metadata": {},
   "source": [
    "### Number CPUs\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa6d999a-788a-4782-a638-0236fd81176b",
   "metadata": {},
   "source": [
    "The number of CPUs allocated to running the task. NOTE:(Should be set/optimized per job scheduler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8676c4b-5601-4e16-b176-f95cc4d2e187",
   "metadata": {},
   "outputs": [],
   "source": [
    "%set_env CPU=40"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d0f84e-c5cc-4f37-9a35-0e9baa56e22b",
   "metadata": {},
   "source": [
    "### Gene Finding Tool\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "092af61a-4efc-41a9-8c59-c50769d30d53",
   "metadata": {},
   "source": [
    "Tool for `antiSMASH` to use to identify and assemble genes, we recommend using prodigal for this purpose."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3801cc8c-6b50-453d-816e-bfa7e426aba2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%set_env GENETOOL=progigal-m"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c73b9e8-0a4e-4596-b7b2-dcd5fa877e43",
   "metadata": {},
   "source": [
    "### Minimum Length \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95bff71d-c00f-404a-91ee-2d7324be4b20",
   "metadata": {},
   "source": [
    "Minimum length of read to be analyzed.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce43f452-9326-4864-8a64-50f2c35a98e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%set_env MINLENGTH=5000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96260cdb-cfc4-4b17-b686-6aba52f13c64",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Data Loading\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9268647b-e16e-45b1-99e5-e31f134daabe",
   "metadata": {},
   "source": [
    "## Download Fasta File Containing Sample\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8820d33-fdc2-45ac-96a0-c413c2d41bb9",
   "metadata": {},
   "source": [
    "Download the demonstration sample from New Atlantis S3 resources. \n",
    "\n",
    "NOTE: additional tests required to ensure s3 is intalled on JHUB and will always work. K+T task."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e32bb103-9ff6-48c6-860f-0656fdd2a550",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "aws s3 cp s3://newatlantis-case-studies/SOLA-samples/ERR2604088 ${WORKDIR}/data/sola/${SAMPLE}* --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55c51227-1cbf-4cb1-ab44-e3904307bdec",
   "metadata": {},
   "source": [
    "## Download Benchmarked Gene Cluster Families (GCF) models from the MIBiG database to be used for refernce in the following analysis.\n",
    "---\n",
    "NOTE: Please provide more info what these are and name MIBiG by name in the input data section of this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e507c7c-d98a-4ced-920b-9c0b73a58f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "aws s3 cp s3://newatlantis-case-studies/mibig_gcf_models ${WORKDIR}/data/mibig_gcf_models --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4f64e2f-e7ea-472d-9942-2010e2dcb193",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Execution of Tool \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "358612e9-411d-4fba-ac26-719e4c9d9cba",
   "metadata": {},
   "source": [
    "The best way to run `antiSMASH` is using the shell script that is provided in the src directory. This file handles runs the tool in the proper docker container and then outputs the results to the output directories specified in the previous section. View `run_antismash.sh` for more detail on how the tool is run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bea7fc90-41cb-42b9-99b4-5e07f73f7383",
   "metadata": {},
   "source": [
    "## Execute antiSMASH shell script\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6fb48725-d14d-4d5c-819e-28cd0583c13b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing Jupyter notebook from /home/jovyan/shared/Active_Projects/Templates/src/NB_utility.ipynb\n",
      "Hello World!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "SCAFFOLDS=$(ls ${WORKDIR}/data/sola/ERR*/output/scaffolds.fasta)\n",
    "for SCAFFOLD in ${SCAFFOLDS}; do\n",
    "  SAMPLE_NAME=$(echo \"${SCAFFOLD}\" | sed \"s/.*\\(ERR[0-9]\\+\\)\\/output.*/\\1/\");\n",
    "  OUTPUT_DIR=\"${WORKDIR}/outputs/antismash/${SAMPLE_NAME}\";\n",
    "  \"${REPO}\"/execution/src/run_antismash.sh \"${SCAFFOLD}\" \"${OUTPUT_DIR}\" \\\n",
    "  --cpus ${CPU} \\\n",
    "  --genefinding-tool ${GENETOOL} \\\n",
    "  --taxon ${TAXON} \\\n",
    "  --allow-long-headers \\\n",
    "  --minlength ${MINLENGTH};\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f2e6cd1-4c15-447a-8cf9-83cd7d58cdba",
   "metadata": {},
   "source": [
    "NOTE: Add any necessary print statements to shell script to ensure its working, though docker handles this well.\n",
    "Add return to shell script that prints to confirm a succesfull run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "535a420e-919a-47ba-a8b5-15e2a0733e54",
   "metadata": {},
   "source": [
    "---\n",
    "# Conclusion\n",
    "---\n",
    "Include any final parting thoughts in this section.\n",
    "This section may also incude:\n",
    "- Common mistakes and fixes. \n",
    "- Debugging tips.\n",
    "- Contact for the author.\n",
    "- Any other information you would like to include"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56339f87-34a1-4fad-a584-9ebfa2be866b",
   "metadata": {},
   "source": [
    "# Write what you want in the above."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9825cda-38b4-4a11-9a35-315f96dfa27f",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3fc821a5-1d24-4899-ae87-24aa86d05c60",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c532b35d-5cc4-48b9-a09f-2f09ebe08d27",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c8001bd4-2ad7-4c32-8117-cd5d5a050bd0",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "505e23f4-f056-4803-b336-973de2b5a061",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
