{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "49c8f402-cc0c-436f-8898-06ecc316f51a",
   "metadata": {},
   "source": [
    "## **Bioprospecting analysis 1.0**\n",
    "\n",
    "#### This notebook integrates the metagenomic bioprospecting analysis  1.0.  \n",
    "#### The tasks are organized in the following sections:  \n",
    "#### [**0. Set env**](#section0)\n",
    "#### [**1. Identify of BGCs**](#section1)\n",
    "#### [**2. Compute BGC class abundances**](#section2)\n",
    "#### [**3. Taxonomic annotation of BGCs**](#section3)\n",
    "#### [**4. Map BGCs onto reference Gene Cluster Families (GCFs)**](#section4)\n",
    "#### [**5. Compute BGCs diversity, functionally annotate, and assess novelty**]((#section5)\n",
    "\n",
    "#### Dependencies to run this notebook (outside the tools we provide):  \n",
    "#### [aws cli](https://aws.amazon.com/cli/)  \n",
    "#### [bash and R kernels](https://evodify.com/python-r-bash-jupyter-notebook/)  \n",
    "#### [Docker](https://www.docker.com/)\n",
    "#### [RSQLite R library](https://cran.r-project.org/web/packages/RSQLite/index.html)\n",
    "#### [tidyverse R library](https://www.tidyverse.org/)\n",
    "#### [seqtk](https://github.com/lh3/seqtk)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cbc2398-db3e-4ec2-9b48-848bc6c832b8",
   "metadata": {},
   "source": [
    "<a id='section0'></a>\n",
    "**0. Set env**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1357053d-48e1-445e-a0de-e42f0da41828",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: WORKDIR=workdir\n",
      "env: REPO=/home/epereira/workspace/dev/new_atlantis/repos/bioprospecting\n",
      "env: seqtk=/nfs/bin/seqtk/seqtk\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython\n",
    "%set_env WORKDIR=workdir\n",
    "%set_env REPO=/home/epereira/workspace/dev/new_atlantis/repos/bioprospecting\n",
    "%set_env seqtk=/nfs/bin/seqtk/seqtk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fc7003a2-c4e3-489f-b171-d1920dbd38df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir -p ${WORKDIR}/data/sola\n",
    "mkdir -p ${WORKDIR}/outputs/antismash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed3a9241-fdfe-4d2d-b5ee-c5537507ad63",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id='section1'></a>\n",
    "**1. Identification of BGCs**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c18d7ac1-ac70-4626-b97e-0610b53e5d61",
   "metadata": {},
   "source": [
    "We will be using the [SOLA metagenomic dataset](https://www.nature.com/articles/s41396-018-0158-1), already assembled with [VEBA](https://github.com/jolespin/veba).\n",
    "Letâ€™s first get the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "378b0f1c-7cb4-47b0-815d-fb8ac3723414",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# aws s3 cp s3://newatlantis-case-studies/SOLA-samples/ ${WORKDIR}/data/sola --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54a468a9-ac35-4c38-967b-6e68bb31bef8",
   "metadata": {},
   "source": [
    "This dataset contains the assembled scaffolds (\\*.fasta) and the mapping files (\\*.bam)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "37e84e70-3389-49d3-9fad-9049dcb70132",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "workdir/data/sola/ERR2604071/output:\n",
      "featurecounts.tsv.gz\n",
      "mapped.sorted.bam\n",
      "mapped.sorted.bam.bai\n",
      "scaffolds.fasta\n",
      "scaffolds.fasta.1.bt2\n",
      "scaffolds.fasta.2.bt2\n",
      "scaffolds.fasta.3.bt2\n",
      "scaffolds.fasta.4.bt2\n",
      "scaffolds.fasta.rev.1.bt2\n",
      "scaffolds.fasta.rev.2.bt2\n",
      "scaffolds.fasta.saf\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "ls ${WORKDIR}/data/sola/ERR*/output | head -12"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdcece54-3f7f-48fc-8bea-cd9d784a00b0",
   "metadata": {},
   "source": [
    "Now that we have the data, let's run [antisMASH](https://github.com/antismash/antismash) to identify the BGC sequences.  \n",
    "For this we will be using our wrap script [run_antismash](https://github.com/pereiramemo/bioprospecting/blob/main/run_scripts/run_antismash.sh), which runs a containerized version 6.0.0 of antiSMASH.  \n",
    "Note that there is version 7.0.0 available, but for compatibility purposes in downstream analysis, we'll use this version for now.\n",
    "Since we are using a wrap script to run a containerized version of antiSMASH, we have to use the fist two positional parameters as the input and output folders, respectively.  \n",
    "To see the help we run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "04c291c2-f727-4198-9941-3cf538950c62",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "########### antiSMASH 6.0.0 #############\n",
      "\n",
      "usage: antismash [-h] [options ..] sequence\n",
      "\n",
      "\n",
      "arguments:\n",
      "  SEQUENCE  GenBank/EMBL/FASTA file(s) containing DNA.\n",
      "\n",
      "--------\n",
      "Options\n",
      "--------\n",
      "-h, --help              Show this help text.\n",
      "--help-showall          Show full lists of arguments on this help text.\n",
      "-c CPUS, --cpus CPUS    How many CPUs to use in parallel. (default: 48)\n",
      "\n",
      "Basic analysis options:\n",
      "\n",
      "  --taxon {bacteria,fungi}\n",
      "                        Taxonomic classification of input sequence. (default:\n",
      "                        bacteria)\n",
      "\n",
      "Additional analysis:\n",
      "\n",
      "  --fullhmmer           Run a whole-genome HMMer analysis.\n",
      "  --cassis              Motif based prediction of SM gene cluster regions.\n",
      "  --clusterhmmer        Run a cluster-limited HMMer analysis.\n",
      "  --tigrfam             Annotate clusters using TIGRFam profiles.\n",
      "  --smcog-trees         Generate phylogenetic trees of sec. met. cluster\n",
      "                        orthologous groups.\n",
      "  --tta-threshold TTA_THRESHOLD\n",
      "                        Lowest GC content to annotate TTA codons at (default:\n",
      "                        0.65).\n",
      "  --cb-general          Compare identified clusters against a database of\n",
      "                        antiSMASH-predicted clusters.\n",
      "  --cb-subclusters      Compare identified clusters against known subclusters\n",
      "                        responsible for synthesising precursors.\n",
      "  --cb-knownclusters    Compare identified clusters against known gene\n",
      "                        clusters from the MIBiG database.\n",
      "  --asf                 Run active site finder analysis.\n",
      "  --pfam2go             Run Pfam to Gene Ontology mapping module.\n",
      "  --rre                 Run RREFinder precision mode on all RiPP gene\n",
      "                        clusters.\n",
      "  --cc-mibig            Run a comparison against the MIBiG dataset\n",
      "\n",
      "Output options:\n",
      "\n",
      "  --output-dir OUTPUT_DIR\n",
      "                        Directory to write results to.\n",
      "  --output-basename OUTPUT_BASENAME\n",
      "                        Base filename to use for output files within the\n",
      "                        output directory.\n",
      "  --html-title HTML_TITLE\n",
      "                        Custom title for the HTML output page (default is\n",
      "                        input filename).\n",
      "  --html-description HTML_DESCRIPTION\n",
      "                        Custom description to add to the output.\n",
      "  --html-start-compact  Use compact view by default for overview page.\n",
      "\n",
      "Gene finding options (ignored when ORFs are annotated):\n",
      "\n",
      "  --genefinding-tool {glimmerhmm,prodigal,prodigal-m,none,error}\n",
      "                        Specify algorithm used for gene finding: GlimmerHMM,\n",
      "                        Prodigal, Prodigal Metagenomic/Anonymous mode, or\n",
      "                        none. The 'error' option will raise an error if\n",
      "                        genefinding is attempted. The 'none' option will not\n",
      "                        run genefinding. (default: error).\n",
      "  --genefinding-gff3 GFF3_FILE\n",
      "                        Specify GFF3 file to extract features from.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "\"${REPO}\"/run_scripts/run_antismash.sh . . --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37d3bfdb-7eb1-40b4-920e-1be67935214b",
   "metadata": {},
   "source": [
    "Let's run antiSMASH on the SOLA dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "53eab970-dfaa-4d2f-a483-0b62306e8a88",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR2604071\n",
      "ERR2604073\n",
      "ERR2604074\n",
      "ERR2604075\n",
      "ERR2604076\n",
      "ERR2604077\n",
      "ERR2604078\n",
      "ERR2604079\n",
      "ERR2604080\n",
      "ERR2604081\n",
      "ERR2604082\n",
      "ERR2604083\n",
      "ERR2604084\n",
      "ERR2604085\n",
      "ERR2604086\n",
      "ERR2604087\n",
      "ERR2604088\n",
      "ERR2604090\n",
      "ERR2604091\n",
      "ERR2604092\n",
      "ERR2604093\n",
      "ERR2604094\n",
      "ERR2604095\n",
      "ERR2604096\n",
      "ERR2604097\n",
      "ERR2604098\n",
      "ERR2604099\n",
      "ERR2604100\n",
      "ERR2604101\n",
      "ERR2604102\n",
      "ERR2604103\n",
      "ERR2604104\n",
      "ERR2604105\n",
      "ERR2604106\n",
      "ERR2604107\n",
      "ERR2604108\n",
      "ERR2604109\n",
      "ERR2604110\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "SCAFFOLDS=$(ls ${WORKDIR}/data/sola/ERR*/output/scaffolds.fasta)\n",
    "for SCAFFOLD in ${SCAFFOLDS}; do\n",
    "\n",
    "  SAMPLE_NAME=$(echo \"${SCAFFOLD}\" | sed \"s/.*\\(ERR[0-9]\\+\\)\\/output.*/\\1/\");\n",
    "  OUTPUT_DIR=\"${WORKDIR}/outputs/antismash/${SAMPLE_NAME}\";\n",
    "  echo \"${SAMPLE_NAME}\";\n",
    " \n",
    "  \"${REPO}\"/run_scripts/run_antismash.sh \"${SCAFFOLD}\" \"${OUTPUT_DIR}\" \\\n",
    "  --cpus 40 \\\n",
    "  --genefinding-tool prodigal-m \\\n",
    "  --taxon bacteria \\\n",
    "  --allow-long-headers \\\n",
    "  --minlength 5000;\n",
    "\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "335ff3ab-7de8-403a-a6ff-f7f193dcdf29",
   "metadata": {},
   "source": [
    "The annotated BGC sequences can be found in `${WORKDIR}/outputs/antismash/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c5f46a84-8b51-4879-94da-c50e657f9fa6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR2604071\n",
      "ERR2604073\n",
      "ERR2604074\n",
      "ERR2604075\n",
      "ERR2604076\n",
      "ERR2604077\n",
      "ERR2604078\n",
      "ERR2604079\n",
      "ERR2604080\n",
      "ERR2604081\n",
      "ERR2604082\n",
      "ERR2604083\n",
      "ERR2604084\n",
      "ERR2604085\n",
      "ERR2604086\n",
      "ERR2604087\n",
      "ERR2604088\n",
      "ERR2604090\n",
      "ERR2604091\n",
      "ERR2604092\n",
      "ERR2604093\n",
      "ERR2604094\n",
      "ERR2604095\n",
      "ERR2604096\n",
      "ERR2604097\n",
      "ERR2604098\n",
      "ERR2604099\n",
      "ERR2604100\n",
      "ERR2604101\n",
      "ERR2604102\n",
      "ERR2604103\n",
      "ERR2604104\n",
      "ERR2604105\n",
      "ERR2604106\n",
      "ERR2604107\n",
      "ERR2604108\n",
      "ERR2604109\n",
      "ERR2604110\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "ls ${WORKDIR}/outputs/antismash/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6635411a-d5af-41f0-b4c5-f8d5bcc06215",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id='section2'></a>\n",
    "**2. Compute BGC class abundances**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d9598fa-b705-4b92-8df4-1a626bf8715e",
   "metadata": {},
   "source": [
    "To do this, let's first create a directory where we will put the concatenated BGC gbk and the bam mapping files.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4701455e-5c0a-4ce3-92ef-b723b67ea684",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# mkdir ${WORKDIR}/outputs/bgc_abund\n",
    "\n",
    "ls -d ${WORKDIR}/outputs/antismash/ERR* | \\\n",
    "while read LINE; do\n",
    "  SAMPLE=$(basename \"${LINE}\")\n",
    "  cat \"${LINE}\"/scaffolds/\"${SAMPLE}\"*.region*.gbk > \"${WORKDIR}\"/outputs/bgc_abund/\"${SAMPLE}\".gbk\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a851173-72f9-498d-9347-869a7a0c707b",
   "metadata": {
    "tags": []
   },
   "source": [
    "Now we can run our custom scripts [get_cov.py](https://github.com/pereiramemo/bioprospecting/blob/main/aux_scripts/get_cov.py).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d79072c7-5b4e-4b7b-8e32-b862384f6f5e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "ls \"${WORKDIR}\"/outputs/bgc_abund/*.gbk | \\\n",
    "while read LINE; do\n",
    "\n",
    "  SAMPLE=$(basename \"${LINE}\" .gbk)\n",
    "  \n",
    "  \"${REPO}\"/aux_scripts/get_cov.py \\\n",
    "  --input_gbk \"${WORKDIR}\"/outputs/bgc_abund/\"${SAMPLE}\".gbk \\\n",
    "  --input_bam \"${WORKDIR}\"/data/sola/\"${SAMPLE}\"/output/mapped.sorted.bam \\\n",
    "  --sample_name \"${SAMPLE}\" \\\n",
    "  --output_tsv \"${WORKDIR}\"/outputs/bgc_abund/\"${SAMPLE}\".tsv\n",
    "\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "62dca4df-ef65-4008-9f46-54c8a0c35446",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR2604071__k119_144067\tphosphonate\tTrue\t0\t6932\t1.0418349682631276\tERR2604071\t\n",
      "ERR2604071__k119_154594\tbetalactone\tTrue\t0\t11920\t1.014010067114094\tERR2604071\t\n",
      "ERR2604071__k119_29665\tresorcinol\tTrue\t0\t6522\t1.0111928856179087\tERR2604071\t\n",
      "ERR2604071__k119_29813\tterpene\tTrue\t0\t12578\t1.02146605183654\tERR2604071\t\n",
      "ERR2604071__k119_50764\tterpene\tTrue\t0\t6293\t1.034164945177181\tERR2604071\t\n",
      "ERR2604071__k119_63513\tterpene\tTrue\t0\t10997\t1.0230972083295444\tERR2604071\t\n",
      "ERR2604071__k119_65188\tterpene\tTrue\t0\t13915\t1.0409629895795904\tERR2604071\t\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head \"${WORKDIR}\"/outputs/bgc_abund/ERR2604071.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f08ec118-e216-4384-a081-747e57988694",
   "metadata": {},
   "source": [
    "We obtain a tsv file with the following fields: `scafflod id, BGC class, edge, start, end, coverage, sample`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b8ef9b4-d5ea-4a31-aa10-0812970e5f79",
   "metadata": {},
   "source": [
    "Let's concatenate all these files into one tsv table (sample, bgc class, coverage):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d3d5dcd2-2c3f-4ff3-a4e3-75cdd34c7be3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR2604071\tphosphonate\t1.0418349682631276\n",
      "ERR2604071\tbetalactone\t1.014010067114094\n",
      "ERR2604071\tresorcinol\t1.0111928856179087\n",
      "ERR2604071\tterpene\t1.02146605183654\n",
      "ERR2604071\tterpene\t1.034164945177181\n",
      "ERR2604071\tterpene\t1.0230972083295444\n",
      "ERR2604071\tterpene\t1.0409629895795904\n",
      "ERR2604073\tterpene\t1.0692621095518333\n",
      "ERR2604073\tbetalactone\t1.0658359680119025\n",
      "ERR2604073\thserlactone\t1.0689745607133492\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cat \"${WORKDIR}\"/outputs/bgc_abund/ERR*.tsv | awk -v OFS=\"\\t\" '{print $7,$2,$6}' > \"${WORKDIR}\"/outputs/bgc_abund/bgc_abund.tsv\n",
    "head \"${WORKDIR}\"/outputs/bgc_abund/bgc_abund.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "718abbac-1fa2-4051-b6aa-6d53ba55fae2",
   "metadata": {},
   "source": [
    "<a id='section3'></a>\n",
    "**3. Taxonomic annotation of BGCs**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9859a4d1-ceb1-40ea-b0d7-247c38a15091",
   "metadata": {
    "tags": []
   },
   "source": [
    "For this we will use the [MMseqs taxonomy](https://github.com/soedinglab/MMseqs2), for which we have containerized and created the corresponding run script [run_mmseqs_taxonomy.sh](https://github.com/pereiramemo/bioprospecting/blob/main/run_scripts/run_mmseqs_taxonomy.sh).  \n",
    "As a reference we will be using the [UniProtKB/Swiss-Prot](https://www.uniprot.org/uniprotkb?facets=reviewed%3Atrue&query=%2A). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6411e68c-3a43-4dbe-b68a-d0492b245784",
   "metadata": {},
   "source": [
    "Before doing anything, we have to organize the data and select the scaffolds in which a BGC was annotated:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7747d8ad-bd66-41f5-8743-a44fb69d01bc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# mkdir \"${WORKDIR}\"/outputs/bgc_taxa\n",
    "\n",
    "cat \"${WORKDIR}\"/outputs/bgc_abund/ERR*.tsv | cut -f1 | sort | uniq > \"${WORKDIR}\"/outputs/bgc_taxa/ids.txt\n",
    "cat \"${WORKDIR}\"/data/sola/ERR*/output/scaffolds.fasta >  \"${WORKDIR}\"/outputs/bgc_taxa/all.fasta\n",
    "\n",
    "\"${seqtk}\" subseq \\\n",
    "\"${WORKDIR}\"/outputs/bgc_taxa/all.fasta \\\n",
    "\"${WORKDIR}\"/outputs/bgc_taxa/ids.txt > \\\n",
    "\"${WORKDIR}\"/outputs/bgc_taxa/bgc.fasta\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be2fe24-7b34-4241-ad0e-08d4a314307e",
   "metadata": {},
   "source": [
    "Now we can perform the taxonomic annotation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5b5a8626-a7b6-48ac-8d64-8e7c9151e185",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "\"${REPO}\"/run_scripts/run_mmseqs_taxonomy.sh \\\n",
    "\"${WORKDIR}\"/outputs/bgc_taxa/bgc.fasta \\\n",
    "\"${WORKDIR}\"/outputs/bgc_taxa/bgc_taxa_annot \\\n",
    "--threads 40 \\\n",
    "--tax-lineage 1 \\\n",
    "-v 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2e4442d5-fdbc-4cd5-ae38-a03f27d62e89",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR2604090__k119_185408\t1890424\torder\tSynechococcales\t26\t25\t18\t0.710\t-_cellular organisms;d_Bacteria;-_Terrabacteria group;-_Cyanobacteriota/Melainabacteria group;p_Cyanobacteriota;c_Cyanophyceae;o_Synechococcales\n",
      "ERR2604090__k119_153928\t2\tsuperkingdom\tBacteria\t9\t8\t6\t0.750\t-_cellular organisms;d_Bacteria\n",
      "ERR2604090__k119_123795\t31989\tfamily\tParacoccaceae\t4\t4\t2\t0.500\t-_cellular organisms;d_Bacteria;p_Pseudomonadota;c_Alphaproteobacteria;o_Rhodobacterales;f_Paracoccaceae\n",
      "ERR2604090__k119_131898\t1236\tclass\tGammaproteobacteria\t7\t7\t5\t0.800\t-_cellular organisms;d_Bacteria;p_Pseudomonadota;c_Gammaproteobacteria\n",
      "ERR2604090__k119_155472\t1224\tphylum\tPseudomonadota\t10\t10\t7\t0.670\t-_cellular organisms;d_Bacteria;p_Pseudomonadota\n",
      "ERR2604090__k119_52769\t131567\tno rank\tcellular organisms\t3\t3\t3\t1.000\t-_cellular organisms\n",
      "ERR2604090__k119_148185\t2\tsuperkingdom\tBacteria\t5\t5\t5\t1.000\t-_cellular organisms;d_Bacteria\n",
      "ERR2604090__k119_84479\t1224\tphylum\tPseudomonadota\t9\t9\t8\t0.880\t-_cellular organisms;d_Bacteria;p_Pseudomonadota\n",
      "ERR2604090__k119_203511\t31989\tfamily\tParacoccaceae\t40\t39\t21\t0.570\t-_cellular organisms;d_Bacteria;p_Pseudomonadota;c_Alphaproteobacteria;o_Rhodobacterales;f_Paracoccaceae\n",
      "ERR2604090__k119_195712\t49546\tfamily\tFlavobacteriaceae\t3\t3\t2\t0.670\t-_cellular organisms;d_Bacteria;-_FCB group;-_Bacteroidota/Chlorobiota group;p_Bacteroidota;c_Flavobacteriia;o_Flavobacteriales;f_Flavobacteriaceae\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head \"${WORKDIR}\"/outputs/bgc_taxa/bgc_taxa_annot_lca.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2cf6c94-38e7-48d4-9269-9b3704e19fc2",
   "metadata": {},
   "source": [
    "<a id='section4'></a>\n",
    "**4. Mapping BGCs onto reference Gene Cluster Families (GCFs)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c42888a2-1139-4204-aab5-5cbd4621af5c",
   "metadata": {},
   "source": [
    "We will use [BiG-SLICE](https://github.com/medema-group/bigslice) to map the identified BGCs onto the GCFs of [BiG-FAM database](https://bigfam.bioinformatics.nl/).  \n",
    "This tool requires the [dataset.tsv and taxonomy files](https://github.com/medema-group/bigslice/wiki/Input-folder). First the `dataset.tsv` with the following commands:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "33e2fc96-927e-4f41-aa35-82f57bd1be81",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "ls -d \"${WORKDIR}/outputs/antismash/\"ERR* | \\\n",
    "while read LINE; do\n",
    "\n",
    "  DATASET=$(basename $(ls -d ${LINE}))\n",
    "  PATH2DATASET=$(basename $(dirname ${LINE}))\"/\"\n",
    "  echo -e \"${DATASET}\\t./\\ttaxonomy/${DATASET}_taxonomy.tsv\\tdataset_${DATASET}\"\n",
    "\n",
    "done > \"${WORKDIR}/outputs/antismash/datasets.tsv\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a02e0ebb-964c-47c8-aa9f-7a29c9ccfc3f",
   "metadata": {},
   "source": [
    "Then, the `taxonomy.tsv` files utilizing our auxiliary script [create_taxonomy.py](https://github.com/pereiramemo/bioprospecting/blob/main/aux_scripts/create_taxonomy.py)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f8b00597-3a67-4d9c-92c3-3983ecc2d514",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "\"${REPO}\"/aux_scripts/create_taxonomy.py \\\n",
    "--input_tsv \"${WORKDIR}\"/outputs/bgc_taxa/bgc_taxa_annot_lca.tsv \\\n",
    "--output_dir \"${WORKDIR}/outputs/antismash/taxonomy\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0abda19-67ee-435b-99f0-fecf7c21dc15",
   "metadata": {
    "tags": []
   },
   "source": [
    "Now let's download the [BiG-FAM database](https://bigfam.bioinformatics.nl/) (this can take some time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c7f84a1e-d112-46fc-8feb-63124f039bc9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# wget http://bioinformatics.nl/~kauts001/ltr/bigslice/paper_data/data/full_run_result.zip --directory-prefix  ${WORKDIR}/data/\n",
    "# unzip ${WORKDIR}/data/full_run_result.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8681a31f-63dc-4561-956d-1aef2f1277aa",
   "metadata": {
    "tags": []
   },
   "source": [
    "And run [BiG-SLICE](https://github.com/medema-group/bigslice) using our containerized version:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bcefcf05-1e26-4fe3-a0d9-786edcf1506b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pid 103's current affinity list: 0-47\n",
      "pid 103's new affinity list: 47\n",
      "pid 104's current affinity list: 0-47\n",
      "pid 104's new affinity list: 46\n",
      "pid 105's current affinity list: 0-47\n",
      "pid 105's new affinity list: 45\n",
      "pid 106's current affinity list: 0-47\n",
      "pid 106's new affinity list: 44\n",
      "pid 107's current affinity list: 0-47\n",
      "pid 107's new affinity list: 43\n",
      "pid 108's current affinity list: 0-47\n",
      "pid 108's new affinity list: 42\n",
      "pid 109's current affinity list: 0-47\n",
      "pid 109's new affinity list: 41\n",
      "pid 110's current affinity list: 0-47\n",
      "pid 110's new affinity list: 40\n",
      "pid 111's current affinity list: 0-47\n",
      "pid 111's new affinity list: 39\n",
      "pid 112's current affinity list: 0-47\n",
      "pid 112's new affinity list: 38\n",
      "pid 113's current affinity list: 0-47\n",
      "pid 113's new affinity list: 37\n",
      "pid 114's current affinity list: 0-47\n",
      "pid 114's new affinity list: 36\n",
      "pid 115's current affinity list: 0-47\n",
      "pid 115's new affinity list: 35\n",
      "pid 116's current affinity list: 0-47\n",
      "pid 116's new affinity list: 34\n",
      "pid 117's current affinity list: 0-47\n",
      "pid 117's new affinity list: 33\n",
      "pid 118's current affinity list: 0-47\n",
      "pid 118's new affinity list: 32\n",
      "pid 119's current affinity list: 0-47\n",
      "pid 119's new affinity list: 31\n",
      "pid 120's current affinity list: 0-47\n",
      "pid 120's new affinity list: 30\n",
      "pid 121's current affinity list: 0-47\n",
      "pid 121's new affinity list: 29\n",
      "pid 122's current affinity list: 0-47\n",
      "pid 122's new affinity list: 28\n",
      "pid 123's current affinity list: 0-47\n",
      "pid 123's new affinity list: 27\n",
      "pid 124's current affinity list: 0-47\n",
      "pid 124's new affinity list: 26\n",
      "pid 125's current affinity list: 0-47\n",
      "pid 125's new affinity list: 25\n",
      "pid 126's current affinity list: 0-47\n",
      "pid 126's new affinity list: 24\n",
      "pid 127's current affinity list: 0-47\n",
      "pid 127's new affinity list: 23\n",
      "pid 128's current affinity list: 0-47\n",
      "pid 128's new affinity list: 22\n",
      "pid 129's current affinity list: 0-47\n",
      "pid 129's new affinity list: 21\n",
      "pid 130's current affinity list: 0-47\n",
      "pid 130's new affinity list: 20\n",
      "pid 131's current affinity list: 0-47\n",
      "pid 131's new affinity list: 19\n",
      "pid 132's current affinity list: 0-47\n",
      "pid 132's new affinity list: 18\n",
      "pid 133's current affinity list: 0-47\n",
      "pid 133's new affinity list: 17\n",
      "pid 134's current affinity list: 0-47\n",
      "pid 134's new affinity list: 16\n",
      "pid 135's current affinity list: 0-47\n",
      "pid 135's new affinity list: 15\n",
      "pid 136's current affinity list: 0-47\n",
      "pid 136's new affinity list: 14\n",
      "pid 137's current affinity list: 0-47\n",
      "pid 137's new affinity list: 13\n",
      "pid 138's current affinity list: 0-47\n",
      "pid 138's new affinity list: 12\n",
      "pid 139's current affinity list: 0-47\n",
      "pid 139's new affinity list: 11\n",
      "pid 140's current affinity list: 0-47\n",
      "pid 140's new affinity list: 10\n",
      "pid 141's current affinity list: 0-47\n",
      "pid 141's new affinity list: 9\n",
      "pid 142's current affinity list: 0-47\n",
      "pid 142's new affinity list: 8\n",
      "pid 1's current affinity list: 0-47\n",
      "pid 1's new affinity list: 8-47\n",
      "Fetching run details...\n",
      "--run_id not specified, using Run#8...\n",
      "Parsing & inserting 745 GBKs...\n",
      "Inserted 745 BGCs!\n",
      "Preparing fasta files for hmmscans...\n",
      "Running hmmscans in parallel...\n",
      "Parsing hmmscans results...\n",
      "Preparing fasta files for subpfam_scans...\n",
      "Running subpfam_scans in parallel...\n",
      "Parsing subpfam_scans results...\n",
      "Extracting features...\n",
      "Assigning GCF membership...\n",
      "Query success! please check results via the output visualization (under 'Reports->View all reports')\n",
      "BiG-SLiCE run complete!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "\"${REPO}\"/run_scripts/run_bigslice.sh query \\\n",
    "\"${WORKDIR}/outputs/antismash/\" \\\n",
    "\"${WORKDIR}/data/full_run_result\" \\\n",
    "--num_threads 40 \\\n",
    "--threshold_pct 0.1 \\\n",
    "--query_name SOLA\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c221b687-8e6c-4a36-90c5-913ae14930c7",
   "metadata": {},
   "source": [
    "We can see the results in the folder `\"${WORKDIR}/data/full_run_result\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c91ef27f-8e39-4544-b833-65c06670f129",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "reports.db\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "ls \"${WORKDIR}/data/full_run_result/reports\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c997e1d-40d1-450e-ac7b-172a3757e2f4",
   "metadata": {},
   "source": [
    "The main result we obtain are the SQLite databases. Although we could access these utilizing the mini web application based on Flask library, we are going to import them into an R environment to have full control of results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0248f43a-6e4d-4c9e-95b0-aac2d4741cfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "library(RSQLite)\n",
    "library(tidyverse)\n",
    "\n",
    "conn_reports_db <- dbConnect(SQLite(), \"workdir/data/full_run_result/reports/1/data.db\")\n",
    "conn_data_db <- dbConnect(SQLite(), \"workdir/data/full_run_result/result/data.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0d819421-810b-4f9a-9754-eaebbda83105",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"bgc\"             \"bgc_class\"       \"bgc_features\"    \"cds\"            \n",
      " [5] \"gcf_membership\"  \"hsp\"             \"hsp_alignment\"   \"hsp_subpfam\"    \n",
      " [9] \"schema\"          \"sqlite_sequence\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "dbListTables(conn_reports_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "aeae54f2-3eff-4d76-848d-dfce2d9cc462",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  gcf_id bgc_id membership_value rank\n",
      "1 244878    378         199.3951    0\n",
      "2 244878    671         199.4237    0\n",
      "3 244878    510         199.3276    0\n",
      "4 244878    201         199.3147    0\n",
      "5 244878    140         199.3254    0\n",
      "6 244878    306         199.2584    0\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "gcf_membership_df <- dbReadTable(conn_reports_db, \"gcf_membership\")\n",
    "head(gcf_membership_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c4286043-362b-4d5f-a61a-444604ef4a44",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 244878\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "gcf_membership_df$gcf_id %>% unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4086c611-cdff-4348-9961-87194dc4fd89",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"bgc\"               \"bgc_class\"         \"bgc_features\"     \n",
      " [4] \"bgc_taxonomy\"      \"cds\"               \"chem_class\"       \n",
      " [7] \"chem_subclass\"     \"chem_subclass_map\" \"clustering\"       \n",
      "[10] \"dataset\"           \"enum_bgc_type\"     \"enum_run_status\"  \n",
      "[13] \"gcf\"               \"gcf_membership\"    \"gcf_models\"       \n",
      "[16] \"hmm\"               \"hmm_db\"            \"hsp\"              \n",
      "[19] \"hsp_alignment\"     \"hsp_subpfam\"       \"run\"              \n",
      "[22] \"run_bgc_status\"    \"run_log\"           \"schema\"           \n",
      "[25] \"sqlite_sequence\"   \"subpfam\"           \"taxon\"            \n",
      "[28] \"taxon_class\"      \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "dbListTables(conn_data_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "609147a8-7fb9-445f-ad54-057100e008e6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   id                       name                   orig_folder\n",
      "1   1                      mibig                     mibig_2.0\n",
      "2   2           isolate_archaeal          ncbi_archaea_genbank\n",
      "3   3 isolate_bacterial_complete ncbi_bacteria_refseq_complete\n",
      "4   4    isolate_bacterial_draft    ncbi_bacteria_refseq_draft\n",
      "5   5             isolate_fungal            ncbi_fungi_genbank\n",
      "6   6               mag_humangut                  almeida_2019\n",
      "7   7                mag_chicken              glendinning_2020\n",
      "8   8                    mag_uba                    parks_2017\n",
      "9   9                 mag_bovine                  stewart_2019\n",
      "10 10                  mag_ocean                    tully_2018\n",
      "                                                                                           description\n",
      "1                                                                             MIBiG 2.0 reference BGCs\n",
      "2                                          BGCs from Archaeal genbanks taken at 2020-02-12 03:22 UTC+1\n",
      "3                BGCs from complete/chromosome-level Bacterial refseqs taken at 2020-02-17 12:15 UTC+1\n",
      "4                    BGCs from contig/scaffold-level Bacterial refseqs taken at 2020-02-17 12:36 UTC+1\n",
      "5                    BGCs from contig/scaffold-level Bacterial refseqs taken at 2020-02-12 03:19 UTC+1\n",
      "6                  BGCs from Metagenome-assembled genomes of Unified Human Gut dataset (Almeida, 2019)\n",
      "7                 BGCs from Metagenome-assembled genomes of Chicken caecum dataset (Glendinning, 2020)\n",
      "8  BGCs from Metagenome-assembled genomes of Uncultivated Bacterial and Archaeal dataset (Parks, 2017)\n",
      "9                        BGCs from Metagenome-assembled genomes of Cow's rumen dataset (Stewart, 2019)\n",
      "10                          BGCs from Metagenome-assembled genomes of TARA ocean dataset (Tully, 2018)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "dbGetQuery(conn_data_db, \"SELECT * FROM dataset LIMIT 10\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7f9bb4-2902-4b86-8171-ee258997aae6",
   "metadata": {},
   "source": [
    "If the dataset is not too large, we can use [BiG-SCAPE](https://github.com/medema-group/BiG-SCAPE) (instead of [BiG-SLICE](https://github.com/medema-group/bigslice)). Although this tool does not scale well, [it has a greater sensitivity compared to BiG-SLICE](https://academic.oup.com/gigascience/article/10/1/giaa154/6092777)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d610e002-b8f6-4ae9-b7f8-12e35fd16113",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: ls [OPTION]... [FILE]...\n",
      "List information about the FILEs (the current directory by default).\n",
      "Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.\n",
      "\n",
      "Mandatory arguments to long options are mandatory for short options too.\n",
      "  -a, --all                  do not ignore entries starting with .\n",
      "  -A, --almost-all           do not list implied . and ..\n",
      "      --author               with -l, print the author of each file\n",
      "  -b, --escape               print C-style escapes for nongraphic characters\n",
      "      --block-size=SIZE      with -l, scale sizes by SIZE when printing them;\n",
      "                               e.g., '--block-size=M'; see SIZE format below\n",
      "  -B, --ignore-backups       do not list implied entries ending with ~\n",
      "  -c                         with -lt: sort by, and show, ctime (time of last\n",
      "                               modification of file status information);\n",
      "                               with -l: show ctime and sort by name;\n",
      "                               otherwise: sort by ctime, newest first\n",
      "  -C                         list entries by columns\n",
      "      --color[=WHEN]         colorize the output; WHEN can be 'always' (default\n",
      "                               if omitted), 'auto', or 'never'; more info below\n",
      "  -d, --directory            list directories themselves, not their contents\n",
      "  -D, --dired                generate output designed for Emacs' dired mode\n",
      "  -f                         do not sort, enable -aU, disable -ls --color\n",
      "  -F, --classify             append indicator (one of */=>@|) to entries\n",
      "      --file-type            likewise, except do not append '*'\n",
      "      --format=WORD          across -x, commas -m, horizontal -x, long -l,\n",
      "                               single-column -1, verbose -l, vertical -C\n",
      "      --full-time            like -l --time-style=full-iso\n",
      "  -g                         like -l, but do not list owner\n",
      "      --group-directories-first\n",
      "                             group directories before files;\n",
      "                               can be augmented with a --sort option, but any\n",
      "                               use of --sort=none (-U) disables grouping\n",
      "  -G, --no-group             in a long listing, don't print group names\n",
      "  -h, --human-readable       with -l and -s, print sizes like 1K 234M 2G etc.\n",
      "      --si                   likewise, but use powers of 1000 not 1024\n",
      "  -H, --dereference-command-line\n",
      "                             follow symbolic links listed on the command line\n",
      "      --dereference-command-line-symlink-to-dir\n",
      "                             follow each command line symbolic link\n",
      "                               that points to a directory\n",
      "      --hide=PATTERN         do not list implied entries matching shell PATTERN\n",
      "                               (overridden by -a or -A)\n",
      "      --hyperlink[=WHEN]     hyperlink file names; WHEN can be 'always'\n",
      "                               (default if omitted), 'auto', or 'never'\n",
      "      --indicator-style=WORD  append indicator with style WORD to entry names:\n",
      "                               none (default), slash (-p),\n",
      "                               file-type (--file-type), classify (-F)\n",
      "  -i, --inode                print the index number of each file\n",
      "  -I, --ignore=PATTERN       do not list implied entries matching shell PATTERN\n",
      "  -k, --kibibytes            default to 1024-byte blocks for disk usage;\n",
      "                               used only with -s and per directory totals\n",
      "  -l                         use a long listing format\n",
      "  -L, --dereference          when showing file information for a symbolic\n",
      "                               link, show information for the file the link\n",
      "                               references rather than for the link itself\n",
      "  -m                         fill width with a comma separated list of entries\n",
      "  -n, --numeric-uid-gid      like -l, but list numeric user and group IDs\n",
      "  -N, --literal              print entry names without quoting\n",
      "  -o                         like -l, but do not list group information\n",
      "  -p, --indicator-style=slash\n",
      "                             append / indicator to directories\n",
      "  -q, --hide-control-chars   print ? instead of nongraphic characters\n",
      "      --show-control-chars   show nongraphic characters as-is (the default,\n",
      "                               unless program is 'ls' and output is a terminal)\n",
      "  -Q, --quote-name           enclose entry names in double quotes\n",
      "      --quoting-style=WORD   use quoting style WORD for entry names:\n",
      "                               literal, locale, shell, shell-always,\n",
      "                               shell-escape, shell-escape-always, c, escape\n",
      "                               (overrides QUOTING_STYLE environment variable)\n",
      "  -r, --reverse              reverse order while sorting\n",
      "  -R, --recursive            list subdirectories recursively\n",
      "  -s, --size                 print the allocated size of each file, in blocks\n",
      "  -S                         sort by file size, largest first\n",
      "      --sort=WORD            sort by WORD instead of name: none (-U), size (-S),\n",
      "                               time (-t), version (-v), extension (-X)\n",
      "      --time=WORD            with -l, show time as WORD instead of default\n",
      "                               modification time: atime or access or use (-u);\n",
      "                               ctime or status (-c); also use specified time\n",
      "                               as sort key if --sort=time (newest first)\n",
      "      --time-style=TIME_STYLE  time/date format with -l; see TIME_STYLE below\n",
      "  -t                         sort by modification time, newest first\n",
      "  -T, --tabsize=COLS         assume tab stops at each COLS instead of 8\n",
      "  -u                         with -lt: sort by, and show, access time;\n",
      "                               with -l: show access time and sort by name;\n",
      "                               otherwise: sort by access time, newest first\n",
      "  -U                         do not sort; list entries in directory order\n",
      "  -v                         natural sort of (version) numbers within text\n",
      "  -w, --width=COLS           set output width to COLS.  0 means no limit\n",
      "  -x                         list entries by lines instead of by columns\n",
      "  -X                         sort alphabetically by entry extension\n",
      "  -Z, --context              print any security context of each file\n",
      "  -1                         list one file per line.  Avoid '\\n' with -q or -b\n",
      "      --help     display this help and exit\n",
      "      --version  output version information and exit\n",
      "\n",
      "The SIZE argument is an integer and optional unit (example: 10K is 10*1024).\n",
      "Units are K,M,G,T,P,E,Z,Y (powers of 1024) or KB,MB,... (powers of 1000).\n",
      "\n",
      "The TIME_STYLE argument can be full-iso, long-iso, iso, locale, or +FORMAT.\n",
      "FORMAT is interpreted like in date(1).  If FORMAT is FORMAT1<newline>FORMAT2,\n",
      "then FORMAT1 applies to non-recent files and FORMAT2 to recent files.\n",
      "TIME_STYLE prefixed with 'posix-' takes effect only outside the POSIX locale.\n",
      "Also the TIME_STYLE environment variable sets the default style to use.\n",
      "\n",
      "Using color to distinguish file types is disabled both by default and\n",
      "with --color=never.  With --color=auto, ls emits color codes only when\n",
      "standard output is connected to a terminal.  The LS_COLORS environment\n",
      "variable can change the settings.  Use the dircolors command to set it.\n",
      "\n",
      "Exit status:\n",
      " 0  if OK,\n",
      " 1  if minor problems (e.g., cannot access subdirectory),\n",
      " 2  if serious trouble (e.g., cannot access command-line argument).\n",
      "\n",
      "GNU coreutils online help: <https://www.gnu.org/software/coreutils/>\n",
      "Full documentation at: <https://www.gnu.org/software/coreutils/ls>\n",
      "or available locally via: info '(coreutils) ls invocation'\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ls \"${REPO}\"/run_scripts/run_bigscape.sh . . --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7ebeb317-155a-4d2e-8613-177055bcd212",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# ls \"${WORKDIR}\"/outputs/antismash/ERR2604*/scaffolds/ERR*.region*.gbk | \\\n",
    "# while read LINE; do \n",
    "\n",
    "#   NEW_FILE_NAME=$(basename \"${LINE/.gbk/_formatted.gbk}\")\n",
    "#   OLD_FILE_NAME=$(basename \"${LINE/.gbk/_old.gbk}\")\n",
    "#   DIR_NAME=$(dirname \"${LINE}\")  \n",
    "   \n",
    "#   awk '{ if (NR == 1){print $1,$2} else {print $0}}' \\\n",
    "#   \"${LINE}\" > \"${DIR_NAME}\"/\"${NEW_FILE_NAME}\"\n",
    "  \n",
    "#   mv \"${LINE}\" \"${DIR_NAME}\"/\"${OLD_FILE_NAME}\"\n",
    "\n",
    "# done  \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e53467e9-9b04-4463-b74a-0d469e3186be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604088__k119_116291 18303 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604088__k119_116291' size '18303' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604088__k119_111249 11770 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604088__k119_111249' size '11770' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604088__k119_102369 19680 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604088__k119_102369' size '19680' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_164105 12202 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_164105' size '12202' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_116363 20840 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_116363' size '20840' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_100703 11763 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_100703' size '11763' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_185838 18882 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_185838' size '18882' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_115520 11478 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_115520' size '11478' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_228463 10328 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_228463' size '10328' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_106789 14124 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_106789' size '14124' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_230819 12264 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_230819' size '12264' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_158639 18845 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_158639' size '18845' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_148585 10226 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_148585' size '10226' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604095__k119_186473 13350 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604095__k119_186473' size '13350' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604099__k119_143162 13565 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604099__k119_143162' size '13565' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604099__k119_238206 11080 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604099__k119_238206' size '11080' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604099__k119_128674 11892 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604099__k119_128674' size '11892' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604099__k119_131577 15692 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604099__k119_131577' size '15692' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604099__k119_157044 21017 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604099__k119_157044' size '21017' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604099__k119_160498 17184 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604099__k119_160498' size '17184' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604099__k119_176704 26820 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604099__k119_176704' size '26820' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_216786 14993 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_216786' size '14993' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_196031 10617 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_196031' size '10617' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_211812 20843 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_211812' size '20843' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_147377 15074 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_147377' size '15074' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_137403 12043 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_137403' size '12043' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_168925 15102 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_168925' size '15102' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_212950 14353 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_212950' size '14353' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_106551 28504 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_106551' size '28504' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_182340 14779 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_182340' size '14779' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604079__k119_174650 14215 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604079__k119_174650' size '14215' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604073__k119_132345 13254 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604073__k119_132345' size '13254' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604073__k119_195262 22002 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604073__k119_195262' size '22002' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604073__k119_216347 29625 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604073__k119_216347' size '29625' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604073__k119_164696 11439 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604073__k119_164696' size '11439' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604096__k119_140006 10501 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604096__k119_140006' size '10501' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604096__k119_144686 11646 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604096__k119_144686' size '11646' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_173439 17899 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_173439' size '17899' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_164562 19448 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_164562' size '19448' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_142509 11783 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_142509' size '11783' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_185110 48751 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_185110' size '48751' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_252711 11961 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_252711' size '11961' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_221837 15243 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_221837' size '15243' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_239027 26083 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_239027' size '26083' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604083__k119_228608 11321 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604083__k119_228608' size '11321' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_240826 10951 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_240826' size '10951' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_251344 22002 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_251344' size '22002' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_107176 11898 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_107176' size '11898' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_151174 18169 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_151174' size '18169' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_124848 13424 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_124848' size '13424' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_222013 14678 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_222013' size '14678' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_172768 12185 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_172768' size '12185' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_239052 11644 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_239052' size '11644' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604075__k119_126519 19417 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604075__k119_126519' size '19417' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604106__k119_249846 13811 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604106__k119_249846' size '13811' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604106__k119_377342 11870 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604106__k119_377342' size '11870' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604106__k119_198148 20996 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604106__k119_198148' size '20996' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604106__k119_166093 25646 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604106__k119_166093' size '25646' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604106__k119_320126 14832 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604106__k119_320126' size '14832' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604106__k119_198325 25121 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604106__k119_198325' size '25121' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604106__k119_135178 13789 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604106__k119_135178' size '13789' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604085__k119_149787 21302 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604085__k119_149787' size '21302' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604085__k119_222837 10758 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604085__k119_222837' size '10758' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604097__k119_176336 11085 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604097__k119_176336' size '11085' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604097__k119_195420 11741 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604097__k119_195420' size '11741' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604108__k119_174382 24213 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604108__k119_174382' size '24213' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604108__k119_213323 10921 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604108__k119_213323' size '10921' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604108__k119_164762 14972 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604108__k119_164762' size '14972' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_199988 20633 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_199988' size '20633' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_112081 19168 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_112081' size '19168' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_159275 10528 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_159275' size '10528' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_161985 16313 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_161985' size '16313' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_137106 10749 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_137106' size '10749' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_125391 21206 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_125391' size '21206' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_142356 20909 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_142356' size '20909' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_217538 10712 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_217538' size '10712' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_131476 13355 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_131476' size '13355' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_148273 15869 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_148273' size '15869' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_219226 15323 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_219226' size '15323' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604100__k119_114433 21427 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604100__k119_114433' size '21427' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604071__k119_154594 11920 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604071__k119_154594' size '11920' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604109__k119_104065 14276 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604109__k119_104065' size '14276' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604109__k119_216708 10286 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604109__k119_216708' size '10286' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604109__k119_190405 12511 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604109__k119_190405' size '12511' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604110__k119_113562 21023 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604110__k119_113562' size '21023' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604110__k119_139571 13516 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604110__k119_139571' size '13516' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604110__k119_162021 22371 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604110__k119_162021' size '22371' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604078__k119_192670 20840 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604078__k119_192670' size '20840' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604078__k119_192627 11569 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604078__k119_192627' size '11569' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604078__k119_185166 16226 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604078__k119_185166' size '16226' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604078__k119_196934 13441 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604078__k119_196934' size '13441' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604078__k119_104668 13169 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604078__k119_104668' size '13169' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604074__k119_188186 17416 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604074__k119_188186' size '17416' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604074__k119_114817 12216 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604074__k119_114817' size '12216' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604074__k119_167632 21023 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604074__k119_167632' size '21023' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604074__k119_106647 18540 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604074__k119_106647' size '18540' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_237638 20837 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_237638' size '20837' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_157035 17512 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_157035' size '17512' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_162779 10316 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_162779' size '10316' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_117163 19345 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_117163' size '19345' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_149988 33281 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_149988' size '33281' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_155564 15452 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_155564' size '15452' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_174399 19155 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_174399' size '19155' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_165548 19069 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_165548' size '19069' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_173743 18329 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_173743' size '18329' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_130174 14154 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_130174' size '14154' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604104__k119_195145 20840 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604104__k119_195145' size '20840' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_176533 12827 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_176533' size '12827' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_156961 18717 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_156961' size '18717' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_129823 20837 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_129823' size '20837' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_216473 32818 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_216473' size '32818' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_183351 15565 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_183351' size '15565' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_157573 13662 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_157573' size '13662' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_176649 15310 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_176649' size '15310' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_142270 12480 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_142270' size '12480' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_224752 21053 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_224752' size '21053' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_140730 33282 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_140730' size '33282' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604081__k119_203649 15181 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604081__k119_203649' size '15181' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604077__k119_107912 14331 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604077__k119_107912' size '14331' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604077__k119_140697 19680 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604077__k119_140697' size '19680' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604101__k119_239595 12027 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604101__k119_239595' size '12027' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604101__k119_115232 18801 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604101__k119_115232' size '18801' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604101__k119_171526 32818 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604101__k119_171526' size '32818' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604101__k119_256901 18498 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604101__k119_256901' size '18498' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604101__k119_164063 12767 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604101__k119_164063' size '12767' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604101__k119_230221 15648 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604101__k119_230221' size '15648' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604107__k119_136735 10554 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604107__k119_136735' size '10554' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604107__k119_134500 14119 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604107__k119_134500' size '14119' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604107__k119_273037 11182 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604107__k119_273037' size '11182' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604107__k119_202308 12388 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604107__k119_202308' size '12388' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_120041 20840 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_120041' size '20840' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_154047 12519 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_154047' size '12519' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_178052 14284 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_178052' size '14284' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_164365 12854 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_164365' size '12854' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_173829 15102 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_173829' size '15102' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_103706 17274 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_103706' size '17274' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_171977 16266 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_171977' size '16266' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_168957 14692 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_168957' size '14692' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_147502 21053 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_147502' size '21053' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604093__k119_155548 10323 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604093__k119_155548' size '10323' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604084__k119_235596 13816 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604084__k119_235596' size '13816' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604084__k119_270248 16103 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604084__k119_270248' size '16103' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_249125 22129 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_249125' size '22129' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_241051 20840 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_241051' size '20840' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_209557 23914 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_209557' size '23914' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_267879 20978 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_267879' size '20978' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_198913 18890 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_198913' size '18890' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_103243 17207 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_103243' size '17207' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_101875 16905 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_101875' size '16905' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_215312 11474 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_215312' size '11474' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_103656 16485 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_103656' size '16485' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_275597 11351 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_275597' size '11351' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_231955 18942 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_231955' size '18942' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_245239 21146 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_245239' size '21146' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_145915 13143 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_145915' size '13143' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604080__k119_268532 32818 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604080__k119_268532' size '32818' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604102__k119_185050 10838 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604102__k119_185050' size '10838' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604102__k119_182218 20840 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604102__k119_182218' size '20840' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604102__k119_102356 21197 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604102__k119_102356' size '21197' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604076__k119_137408 12451 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604076__k119_137408' size '12451' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604076__k119_123898 10879 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604076__k119_123898' size '10879' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604076__k119_158569 12053 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604076__k119_158569' size '12053' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604076__k119_162427 16477 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604076__k119_162427' size '16477' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604098__k119_308710 16315 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604098__k119_308710' size '16315' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604098__k119_274359 16900 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604098__k119_274359' size '16900' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604098__k119_234105 25961 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604098__k119_234105' size '25961' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604098__k119_280307 16028 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604098__k119_280307' size '16028' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604098__k119_137774 14513 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604098__k119_137774' size '14513' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604098__k119_310000 21023 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604098__k119_310000' size '21023' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604098__k119_106334 20063 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604098__k119_106334' size '20063' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604103__k119_199992 10339 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604103__k119_199992' size '10339' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604103__k119_136182 10731 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604103__k119_136182' size '10731' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604103__k119_127085 17375 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604103__k119_127085' size '17375' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604103__k119_125276 12045 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604103__k119_125276' size '12045' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604103__k119_101174 13542 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604103__k119_101174' size '13542' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604103__k119_156679 19990 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604103__k119_156679' size '19990' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_185408 20909 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_185408' size '20909' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_148185 12393 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_148185' size '12393' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_153928 13778 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_153928' size '13778' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_203511 21017 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_203511' size '21017' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_151903 21505 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_151903' size '21505' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_168734 20837 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_168734' size '20837' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_126639 10341 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_126639' size '10341' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_103883 16386 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_103883' size '16386' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_199728 20633 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_199728' size '20633' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_181640 11248 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_181640' size '11248' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_224386 10433 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_224386' size '10433' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_206825 21208 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_206825' size '21208' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_118695 17446 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_118695' size '17446' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_155472 11625 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_155472' size '11625' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604090__k119_128614 10861 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604090__k119_128614' size '10861' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604105__k119_210627 12539 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604105__k119_210627' size '12539' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_123130 17314 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_123130' size '17314' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_180031 20423 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_180031' size '20423' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_179612 10993 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_179612' size '10993' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_121784 10891 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_121784' size '10891' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_198442 19341 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_198442' size '19341' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_201034 20439 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_201034' size '20439' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_208174 12829 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_208174' size '12829' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_180136 16375 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_180136' size '16375' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_206204 12134 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_206204' size '12134' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_138835 12084 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_138835' size '12084' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_189824 13498 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_189824' size '13498' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_188866 21020 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_188866' size '21020' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_167424 15102 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_167424' size '15102' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_178589 11575 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_178589' size '11575' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_192110 12611 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_192110' size '12611' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604094__k119_180031 20846 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604094__k119_180031' size '20846' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_188545 12039 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_188545' size '12039' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_109864 21521 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_109864' size '21521' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_187637 14161 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_187637' size '14161' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_151089 15047 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_151089' size '15047' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_146604 10629 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_146604' size '10629' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_145214 21146 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_145214' size '21146' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_114997 12169 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_114997' size '12169' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604092__k119_129312 12145 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604092__k119_129312' size '12145' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604087__k119_206768 19779 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604087__k119_206768' size '19779' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604087__k119_200627 12559 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604087__k119_200627' size '12559' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604087__k119_136530 20282 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604087__k119_136530' size '20282' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604087__k119_136052 25961 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604087__k119_136052' size '25961' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_221518 23220 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_221518' size '23220' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_291904 10853 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_291904' size '10853' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_260765 20014 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_260765' size '20014' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_261756 21415 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_261756' size '21415' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_154220 10400 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_154220' size '10400' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_258165 29408 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_258165' size '29408' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_214593 15088 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_214593' size '15088' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604086__k119_277510 11133 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604086__k119_277510' size '11133' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_124575 14166 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_124575' size '14166' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_169929 25078 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_169929' size '25078' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_164433 15682 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_164433' size '15682' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_138933 20048 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_138933' size '20048' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_181624 16884 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_181624' size '16884' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_119182 10159 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_119182' size '10159' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_178498 21134 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_178498' size '21134' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_174008 29309 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_174008' size '29309' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_144080 20909 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_144080' size '20909' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_145788 16943 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_145788' size '16943' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_168907 10879 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_168907' size '10879' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_170327 21023 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_170327' size '21023' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_145294 12398 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_145294' size '12398' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_116093 20840 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_116093' size '20840' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_184598 47382 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_184598' size '47382' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_183926 16681 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_183926' size '16681' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_132743 13790 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_132743' size '13790' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n",
      "/usr/local/envs/bigscape/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1354: BiopythonParserWarning: Attempting to parse malformed locus line:\n",
      "'LOCUS       ERR2604091__k119_107015 24147 bp    DNA     linear   UNK 01-JAN-1980\\n'\n",
      "Found locus 'ERR2604091__k119_107015' size '24147' residue_type 'DNA'\n",
      "Some fields may be wrong.\n",
      "  BiopythonParserWarning)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "\"${REPO}\"/run_scripts/run_bigscape.sh \\\n",
    "\"${WORKDIR}/outputs/antismash/\" \\\n",
    "\"${WORKDIR}/outputs/bigscape_clust/\" \\\n",
    "--cores 40 > \"${WORKDIR}\"/outputs/bigscape.log\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
