{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "edbe3e02-a1ac-4540-b368-0a8b3d64c2a6",
   "metadata": {},
   "source": [
    "![image](https://cdn.discordapp.com/attachments/996200880351215636/1065002848355631165/New_Atlantis.png) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "321d56b0-ecb4-47e9-b471-8fcee9da8a87",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Execution of the antiSMASH Tool\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad9045a-b22a-4690-8566-cbf08e920b28",
   "metadata": {},
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3feb993-eb12-435d-a51f-856bc6028ea3",
   "metadata": {},
   "source": [
    " \n",
    " This notebook will give a demonstration of running the tool antiSMASH on assembled metagenomic data to identify Biosynthetic Gene Clusters (BGCs) from amongst the reads. `antiSMASH` stands for (antibiotics and Secondary Metabolite Analysis Shell) and is a comprehensive genome mining tool used for the automated identification, annotation, and analysis of secondary metabolite biosynthesis gene clusters in bacterial and fungal genomes. Secondary metabolites are bioactive compounds, often with therapeutic potential, including antibiotics, antifungals and antitumorals.\n",
    "'antiSMASH' has quite a few features, but here we utilize only its ability to identify BGCs It can identify all known types of BGCs such as those responsible for the production of polyketides, non-ribosomal peptides, terpenes, etc.\n",
    "   \n",
    "   \n",
    "The `antiSMASH` tool provides biologists and bioinformaticians with a useful means of deep-diving into genomic data to accelerate the discovery and diversification of biologically active secondary metabolites."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a41f9a9-6c2d-47a9-a063-5e9c5b07e0a5",
   "metadata": {},
   "source": [
    "Markdown cell giving a brief explanation of the tool. Cover what insight is hoped to be generated by the tool and a extremely brief desciption of its methods."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77c9dea-7126-4093-8fb3-ef15b5a6494a",
   "metadata": {},
   "source": [
    "## Input Data \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e028d77b-b270-4032-949f-2d2bd47125e6",
   "metadata": {},
   "source": [
    "The input data are metagenomic samples previously preprocessed and assembled with [VEBA](https://github.com/jolespin/veba), a metagenomic assembly tool. For the purpose of this notebook, we will analyze metagenomes originally from the SOLA dataset. \n",
    "The input consists of:\n",
    "1. **Assembled contigs**: .fasta files.\n",
    "\n",
    "2. **Mapping files** of the assembled contigs in .bam files.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2803ae5-51dc-4a2d-b442-170a8f18412d",
   "metadata": {},
   "source": [
    "## Output Data \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b85d43d-7a3c-4b46-8059-6293d0ec700a",
   "metadata": {},
   "source": [
    "`antiSMASH` outputs an array of data related to the identified secondary metabolite biosynthesis gene clusters (BGCs). The output is highly comprehensive and includes the following:\n",
    "\n",
    "1. **Gene Cluster Annotations:** For each identified BGC, antiSMASH provides an annotation of the genes in the cluster, including the predicted function and potential products.\n",
    "\n",
    "2. **Cluster Border Prediction:** It includes the coordinates of the predicted gene cluster borders in the genome sequence.\n",
    "\n",
    "3. **Biosynthetic Pathway Predictions:** Based on identified enzymatic domains, it predicts potential biosynthetic pathways that could be used by the organism to produce secondary metabolites.\n",
    "\n",
    "4. **Comparative Analysis Data:** The software performs a comparative analysis of the identified gene clusters against known BGCs, yielding similarities and potential novelty.\n",
    "\n",
    "5. **Cluster Networking Data:** antiSMASH can generate data on co-occurring clusters, sharing their potential functional relationships.\n",
    "\n",
    "6. **Data Files:** The software delivers the analysis results in various formats including text-format summary files, GenBank files, SVG images for cluster diagrams, and .gbk files for downstream analysis.\n",
    "\n",
    "7. **Interactive HTML reports:** antiSMASH creates an easy-to-navigate HTML report summarizing all the results of the analysis.\n",
    "\n",
    "8. **Relaxed Cluster Prediction:** In addition to the core areas of BGCs, antiSMASH identifies regions of the genome that could potentially contribute to secondary metabolite production, even if they donâ€™t form a traditional cluster.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad90a79e-33d2-4b4e-bbfe-9987c18b25f4",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Environment\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "493ca348-566c-400f-855e-13e2edb06cf2",
   "metadata": {},
   "source": [
    "This section should point to where the requirements file is store and should be a list of necessary depedencies, install instructions and any auxilliary informantion required to set of the evironment. \n",
    "All requirements files should include the \"import-ipynb\" package in addition to any required to run the tool if you are using a notebook to host source code. This package allows for having notebook based file for storing and documenting utility functions that will be applied later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f21efd60-c031-4586-bb0a-23b552a60482",
   "metadata": {},
   "outputs": [],
   "source": [
    "#install for import-ipynb that will be applied later\n",
    "!pip install import-ipynb -q"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d765ab46-5875-4a9d-97e5-4bcf21631ed8",
   "metadata": {},
   "source": [
    "The following instructions set environment pathing variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0135341f-6059-4f25-aecc-f8e1d27ae306",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: WORKDIR=workdir\n",
      "env: REPO=/home/epereira/workspace/dev/new_atlantis/repos/bioprospecting\n",
      "env: seqtk=/nfs/bin/seqtk/seqtk\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/home/jovyan/shared/Active_Projects/Templates'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext rpy2.ipython\n",
    "%set_env WORKDIR=workdir\n",
    "%set_env REPO=/home/epereira/workspace/dev/new_atlantis/repos/bioprospecting\n",
    "%set_env seqtk=/nfs/bin/seqtk/seqtk\n",
    "\n",
    "import os \n",
    "cwd = os.getcwd()\n",
    "cwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c0d1c43-1e47-4e7f-9ba4-17380d069a68",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING conda.models.version:get_matcher(546): Using .* with relational operator is superfluous and deprecated and will be removed in a future version of conda. Your spec was 1.7.1.*, but conda is ignoring the .* and treating it as 1.7.1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 23.1.0\n",
      "  latest version: 23.9.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=23.9.0\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# All requested packages already installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## conda not functional yet\n",
    "%%bash \n",
    "conda install numpy -c conda-forge\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba91e36d-a9db-4de2-a79e-597db5a0fced",
   "metadata": {},
   "source": [
    "## R Based Dependencies\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e00ac888-f20b-408a-aa81-f656cf927c61",
   "metadata": {},
   "source": [
    "Insert handling and instructions for the above."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f32a969e-f32d-4ff6-bd64-27244614a6c7",
   "metadata": {},
   "source": [
    "---\n",
    "## Import Statements (code)(import ipynb)\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "050d9b6c-a887-43bd-add9-263b913bfb53",
   "metadata": {},
   "outputs": [],
   "source": [
    "import import_ipynb \n",
    "\n",
    "#import all untility functions from utility_func nb \n",
    "from src.NB_utility import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e8a951-1ffa-4af3-a31f-9126426284c6",
   "metadata": {},
   "source": [
    "You may also use a regular Python module to host your source code and import it here. The purpose of using a notebook over a traditional module is for readibility with markdown, more so than for any additional utility. Since they require slightly different handling than importing a python module, examples herein use a utility Notebook. Still do view src/utility_NB.ipynb for documentation instrutions. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8ea48f2-1fd5-4f97-aa2a-872c0b7d7b65",
   "metadata": {},
   "source": [
    "Put all python import statements in a single cell in this section.\n",
    "If any global settings for packages are altered (such as setting matplotlib to inline) do that here in a separate cell immediately after the import statements."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e51905b-74dd-42d9-9b66-5e1e8f56cb60",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Parameters\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e23c198d-fcef-4acf-a53d-2d7af3cc9f07",
   "metadata": {},
   "source": [
    "Add handling for bash flags and r - object oriented parameter approach. parameter object?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3131b0ef-07d3-431b-9138-8c6f4e820496",
   "metadata": {},
   "source": [
    "This section should describe all parameters and setting used to input into the tool. All user supplied arguments should be defined and explained in this section.\n",
    "This section should alternate markdown and code, the markdown explains what a parameter does and what the options are. The first parameter cell should always be paths to data.\n",
    "The first section should contain one cell that only specifies data locations, such that a user only has to edit this cell in order to point the notebook to their data. The next parameter cell should contain paths to any config files or reference databases. The rest of the section should alternate between markdown explaining the parameter followed by code to set the parameter to some default value in either python or env variables. Duplicate the follwoing examples for each paramter used during data cleaning and execution of the tool."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0087c162-c5f3-43d7-a9a1-c74d17b53355",
   "metadata": {},
   "source": [
    "## Input and output data directories \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "197359c5-60ed-48a9-b171-f2e92e9dabc3",
   "metadata": {},
   "source": [
    "These variables are strings that conatin the path to the input and output data directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73a075fa-b2ed-404e-8482-143d5495e01f",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = 'path/to/data'\n",
    "output_dir = 'path/to/data'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e013f57-2555-4060-8be1-29e87ae41a51",
   "metadata": {},
   "source": [
    "This cell sets an environment variable to be used later. This is a useful way to set parameters for shell scripts or bash used later in the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8112d729-1aa1-4c6c-a02a-e6c0ea23b1e0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os \n",
    "\n",
    "os.environ['MESSAGE'] = 'Hello Worlds'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96260cdb-cfc4-4b17-b686-6aba52f13c64",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Data Precleaning (if required) \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8820d33-fdc2-45ac-96a0-c413c2d41bb9",
   "metadata": {},
   "source": [
    "This section aims to demonstrate any modifications done to the data from the original source to the format it the tool needs for processing. Following the instructions in this section, a user should be able to replicate the steps required to take the original data set (or any similar data set) and perform requisite transformations required for the tool. If there is any metadata, augmentations to reference data bases or similar required to run your tool, all those steps should also be included here. \n",
    "This section should follow these steps:\n",
    "- Load data into notebook (pd.read_csv, bash csv, etc)(if required)\n",
    "- Validate the loaded data is the correct format (for user replication) \n",
    "- Perform any other necessary validations (depending on protocol/tool)\n",
    "- Alternate markdown cell explaining manipulation with code cell executing the manipulation.\n",
    "- Mark each step heading with ### to insert into table fo contents \n",
    "\n",
    "If any of these steps are not required for your specific tool, do omit them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4f64e2f-e7ea-472d-9942-2010e2dcb193",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Execution of Tool \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "358612e9-411d-4fba-ac26-719e4c9d9cba",
   "metadata": {},
   "source": [
    "This section aims to demonstrate how to execute the tool and performs a sample run on test data. This portion of the notebook may be fairly code intensive and is the most important part of the notebook. To improve readibility and clarity, most of the verbose code segments should be written as functions in python, as shell scripts stored in the src directory, or scripts in whichever language necessary for the tool you are using.\n",
    "\n",
    "If the step you hope to perform involves more than a couple lines of code, please see the function definition format in the src/NB_utility.ipynb and wrap your code in a function using that format. If you prefer to use bash scripts, wrtie your commands into a shell file, and then execute them in this portion of the notebook. Once your code is wrapped as function in the utility NB, you can inport and run it using the format below.\n",
    "Each step should be separated by a markdown heading with a brief explanation followed by the necessary code.\n",
    "Use the parameter variables definied earlier in the notebook as arguments for functions written here. If there is an input that is not already included in the parameters section, include it there."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bea7fc90-41cb-42b9-99b4-5e07f73f7383",
   "metadata": {},
   "source": [
    "Example function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6fb48725-d14d-4d5c-819e-28cd0583c13b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing Jupyter notebook from /home/jovyan/shared/Active_Projects/Templates/src/NB_utility.ipynb\n",
      "Hello World!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#remember this code was executed earlier \n",
    "import import_ipynb\n",
    "from src.NB_utility import *\n",
    "\n",
    "#now I can execute the hello_world function that was defined in the utility_nb\n",
    "hello_world()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31aa663f-fc33-4af2-b25b-195edeeb5f0a",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Data Post Processing (if required) \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96d262b3-64e7-47ec-bb18-41d1691612aa",
   "metadata": {},
   "source": [
    "## Write to output directory\n",
    "---\n",
    "If the tool does not do it automatically, use this cell to write the output data to the output directory defined in the parameter section."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a95f5b0f-10dc-429d-82a0-e5ce3b55a7a5",
   "metadata": {},
   "source": [
    "This section aims to contain all the code necessary to perform the data cleaning, formatting or analysis that would be performed on the output of this tool. Use the same formatting as previously mentioned in the execution section of the notebook:\n",
    "- Offload long code sections to the src/Utility_NB and import the code \n",
    "- Add validation to catch errors in and irregularities in the data \n",
    "- Alternate code and markdown cells \n",
    "- Include a markdown header for each step using ### to add it to the table of contents\n",
    "- Display data and transformations where necessary. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "111075ec-19f5-4e53-acca-7eacae2ab466",
   "metadata": {},
   "source": [
    "---\n",
    "# Visualization \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdd46e51-4f2e-4811-ae5c-9ceb3f8d7046",
   "metadata": {},
   "source": [
    "If there is a visualization you would like to include here, generate it here.\n",
    "Phrase the code used to generate the visualization as a function in the format mentioned in the execution section of this notebook.\n",
    "Place the function is the utility NB such that it can be reused to generate new visualizations on future data. \n",
    "If the vizualization has additional options and parameters, there is no need to add them to the parameters section, and those parameters can be included into a miniature parameter section  in this section."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "535a420e-919a-47ba-a8b5-15e2a0733e54",
   "metadata": {},
   "source": [
    "---\n",
    "# Conclusion\n",
    "---\n",
    "Include any final parting thoughts in this section.\n",
    "This section may also incude:\n",
    "- Common mistakes and fixes. \n",
    "- Debugging tips.\n",
    "- Contact for the author.\n",
    "- Any other information you would like to include"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56339f87-34a1-4fad-a584-9ebfa2be866b",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c9825cda-38b4-4a11-9a35-315f96dfa27f",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3fc821a5-1d24-4899-ae87-24aa86d05c60",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c532b35d-5cc4-48b9-a09f-2f09ebe08d27",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c8001bd4-2ad7-4c32-8117-cd5d5a050bd0",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "505e23f4-f056-4803-b336-973de2b5a061",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
