{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "534aa71b-cf09-4084-8a3d-5f15ff3abd58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import pysam\n",
    "import csv\n",
    "import pandas as pd\n",
    "\n",
    "def get_coverage(input_dir : str = None, input_bam : str = None, \\\n",
    "                 sample_name : str = None) -> pd.DataFrame:\n",
    "\n",
    "    '''\n",
    "    input_dir:\n",
    "    \ttype: str\n",
    "    \tThis is the output of antiSMASH, where the *.region*.gbk files\n",
    "    \t(i.e., identified BGCs), can be found.\n",
    "\tinput_bam:\n",
    "    \tThese are the bam files representing the alignments of paired end\n",
    "    \treads mapped on the assembled data.\n",
    "\tsample_name:\n",
    "    \tSample name to be used in the output (tsv) table.   \n",
    "\treturns:\n",
    "    \tA data frame with the following columns: \"acc\", \"bgc_class\", \"on_edge\", \"start\", \"end\", \"coverage\", \"sample_name\".\n",
    "\t'''\n",
    " \n",
    "    '''\n",
    "    Start: subfunctions that we will be using in the get_coverage function\n",
    "    '''\n",
    "\n",
    "    def find_files(input_dir : str = None) -> list:\n",
    "        pattern = r\".*\\.region\\d+.gbk\"\n",
    "        matching_files = []\n",
    "        for root, dirs, files in os.walk(input_dir):\n",
    "            for file in files:\n",
    "                if re.search(pattern, file):\n",
    "                    matching_files.append(os.path.join(root, file))\n",
    "        return matching_files\n",
    "\n",
    "    def get_features(record, feature_name, qualifier_name):\n",
    "       for feature in record.features:\n",
    "           if feature.type == feature_name:\n",
    "               return feature.qualifiers[qualifier_name][0]\n",
    "    \n",
    "    def get_feature_location(record, feature_name):\n",
    "       for feature in record.features:\n",
    "          if feature.type == feature_name:\n",
    "             return [int(feature.location.start), int(feature.location.end)]\n",
    "    \n",
    "    def gbk_parser(paths2gbk):\n",
    "        acc2x = dict()   \n",
    "        for input_gbk in paths2gbk:\n",
    "            i = 1\n",
    "            with open(input_gbk, \"r\") as gbk_handle:\n",
    "                record = SeqIO.read(gbk_handle, \"genbank\")\n",
    "                acc = record.annotations['accessions'][0]\n",
    "                bgc_class = get_features(record,\"cand_cluster\", \"product\")\n",
    "                contg_edge = get_features(record,\"cand_cluster\", \"contig_edge\")  \n",
    "                location = get_feature_location(record,\"cand_cluster\")\n",
    "                acc_i = acc + \"-\" + str(i)\n",
    "                i += 1\n",
    "                acc2x[acc_i] = {\"acc\": acc, \"bgc_class\": bgc_class, \"contg_edge\": contg_edge, \"start\":location[0], \"end\":location[1]}\n",
    "        return(acc2x)\n",
    "    \n",
    "    def custom_coverage(input_bam_handle, gbk_parser_out):\n",
    "        coverage_dict = dict()\n",
    "        seq_names = list(gbk_parser_out.keys())\n",
    "        for i in seq_names:\n",
    "            input_bam_cov = input_bam_handle.count_coverage(gbk_parser_out[i][\"acc\"], \\\n",
    "                                                            gbk_parser_out[i][\"start\"], \\\n",
    "                                                            gbk_parser_out[i][\"end\"],\n",
    "                                                            quality_threshold = 0)\n",
    "            n = 0\n",
    "            for j in range (4):\n",
    "                n += len([x for x in input_bam_cov[j] if x != 0])\n",
    "                \n",
    "            coverage_dict[i] = n/(gbk_parser_out[i][\"end\"] - gbk_parser_out[i][\"start\"]) \n",
    "        return(coverage_dict)\n",
    "\n",
    "    '''\n",
    "    End: subfunctions that we will be using in the get_coverage function\n",
    "    '''\n",
    "    \n",
    "    '''\n",
    "    1. Find all input gbk files\n",
    "    '''\n",
    "    \n",
    "    paths2gbk = find_files(input_dir)\n",
    "    \n",
    "    '''\n",
    "    2. Parse GBK\n",
    "    '''\n",
    "\n",
    "    gbk_parser_out = gbk_parser(paths2gbk)\n",
    "   \n",
    "    '''\n",
    "    3. Compute coverage\n",
    "    '''\n",
    "\n",
    "    input_bam_handle = pysam.AlignmentFile(input_bam, \"rb\")\n",
    "    coverage_out = custom_coverage(input_bam_handle, gbk_parser_out)\n",
    "    input_bam_handle.close()\n",
    "\n",
    "    '''\n",
    "    4. Add coverage values to create the output dictionary\n",
    "    '''\n",
    "\n",
    "    output = dict()\n",
    "    for i in gbk_parser_out.keys():\n",
    "        gbk_parser_out[i][\"coverage\"] = coverage_out[i]\n",
    "        gbk_parser_out[i][\"sample\"] = sample_name\n",
    "        output[i] = list(gbk_parser_out[i].values())\n",
    "\n",
    "    '''\n",
    "    5. Convert dictionary to data frame\n",
    "    '''\n",
    "    \n",
    "    column_names = [\"acc\", \"bgc_class\", \"on_edge\", \"start\", \"end\", \"coverage\", \"sample_name\"]\n",
    "    output_df = pd.DataFrame(index = list(output.keys()), columns = column_names)\n",
    "    for i in output.keys():\n",
    "        output_df.loc[i] = output[i]\n",
    "\n",
    "    return(output_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acbbe633-80b9-4c24-b7c4-fbedfea527dc",
   "metadata": {},
   "source": [
    "### Description of get_coverage function"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a5ceebe-5f49-4495-bcc2-e1d2e228e517",
   "metadata": {},
   "source": [
    "This function computes the coverage of all BGCs identified by anitiSMASH. It performs the following tasks: 1) Searches for the \\*region\\*.gbk files within the input folder; 2) Extracts the accession of the contig in which the BGC was identified, and the BGC class, start and end coordinates from the features section; 3) Based on the accession id, and the start and end positions it computes the BGC coverage; 4) It outputs a data frame object with the columns \"acc\", \"bgc_class\", \"on_edge\", \"start\", \"end\", \"coverage\", \"sample_name\", where each row corresponds to a identified BGC. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
